### PR TITLE
[MachineCopyPropagation, Scheduler] Detect and fix suboptimal instruction order to enable optimizations

### PR DESCRIPTION
**Pull request by spaits**
## The issue
In the Coremark becnhmark the following code can be found in core_state.c's core_bench_state function (https://github.com/eembc/coremark/blob/d5fad6bd094899101a4e5fd53af7298160ced6ab/core_state.c#L61):
```c
for (i = 0; i < NUM_CORE_STATES; i++)
{
    crc = crcu32(final_counts[i], crc);
    crc = crcu32(track_counts[i], crc);
}
```
Let's have some code that can be compiled by itself that also reproduces the issue:
```c
void init_var(int *v);
int chain(int c, int n);

void start() {
    int a, b, c;

    init_var(&a);
    init_var(&b);
    init_var(&c);

    int r = chain(b, a);
    r = chain(c, r);
}
```
https://godbolt.org/z/7h8E9nG5q

Clang produces the following assembly for the section between the two function call (on aarch64/arm64):
```asm
bl      _Z5chainii
ldr     w8, [sp, #4]
mov     w1, w0
mov     w0, w8
bl      _Z5chainii
```
While GCC produces this assembly:
```asm
bl      _Z5chainii
mov     w1, w0
ldr     w0, [sp, 28]
bl      _Z5chainii
```
I think we shouldn't move these values around so much.

As you can see gcc does not "shuffle" the values from register to regsiter, but solves the "switching" with only two instruction.

This problem is also present for riscv and it is even worse.
Clang generates:
```asm
call    _Z5chainii
lw      a1, 0(sp)
mv      a2, a0
mv      a0, a1
mv      a1, a2
call    _Z5chainii
```
GCC generates:
```asm
call    _Z5chainii
mv      a1,a0
lw      a0,12(sp)
call    _Z5chainii
```
Also see on godbolt: https://godbolt.org/z/77rncrb3b

The dissasembly of coremark will also have this suboptimal pattern:
```asm
jal	0x8000f596 <crcu32>
lw	a1, 0x24(sp)
mv	a2, a0
mv	a0, a1
mv	a1, a2
jal	0x8000f596 <crcu32>
```

## The cause
The reason for the sub optimal code is the inability of copy propagation to find optimization due to sub optimal order of instructions. The sub optimal order of instructions introduces unnecessary data dependencies between instructions. (Let's say there is data dependency between MI A and MI B if there exists a register unit that is used or defined by both A and B.)

The reason for this data dependency in the examples above is that, the scheduler places loads as early as possible.
This is usually a good thing see https://discourse.llvm.org/t/how-to-copy-propagate-physical-register-introduced-before-ra/74828/4.

After creating the current patch and checking the regression test I found out, that the issue is more general, than the scheduler prioritizing loads. If you look at the test cases you can see that the current state of the patch also enables optimization that have nothing to do with loads, so there are other cases when unnecessary data dependencies block optimizations. (See the changes in llvm/test/CodeGen/RISCV/double-fcmp-strict.ll llvm/test/CodeGen/AArch64/sve-vector-interleave.ll and many more not load related improvements.)

## The current solution
The issue is more generic that sub optimal data dependencies occurring when prioritizing loads, so I think adjusting the scheduler is not enough. Also this affects almost all the targets (see the tests).

I think the proper way to solve this issue is to make the machine copy propagation
"smarter", so in the cases where an unnecessary data dependency(ies) is(are) blocking an optimization
it is recognized and resolved by moving the instruction(s) without changing the semantics.

I have created a new struct that can administer the data dependencies of instructions in a tree.
Also added logic that uses this tree to find unnecessary data dependencies.

For example let's have the following MIR:
```llvm
0. r1 = LOAD sp
1. r2 = COPY r5
2. r3 = ADD r1 r4
3. r1 = COPY r2
```

Let's see what are the dependencies:
Instruction 3 has a dependency on instruction 2 since instruction two uses a value that is in r1 and r1 is re-defined by 3. This means that instruction 2 must precede instruction 3. (This is called an anti dependency.)

Instruction 3 has a dependency on instruction 0. Both of these instructions define r1. instruction 0 must precede instruction 3. (This is called an out dependency.)

And finally instruction 2 has also a dependency on instruction 0, since the value in r1 in the time of execution instruction 2 is put into r1 in instruction 0. (This is called a true dependency.)

From the above, we can deduce, that instruction zero must precede instruction 2 and instruction 2 must precede instruction 3.

We can also observe that instruction 3 has a dependency on instruction 1 (true dependency). Basically in 1 and 3 we are just moving these values around. Could we erase this moving around and instantly do r1 = COPY r5? We can't just do that, since as seen before instruction 2 must precede any redefinition of r1. So in the current case this optimization is blocked by instruction 2.

What if we have switched instruction 2 ad instruction 1 they don't have dependency on each other. Then we will get this:
```llvm
0. r1 = LOAD sp
2. r3 = ADD r1 r4
1. r2 = COPY r5 ; instruction 1 and 2 switched places
3. r1 = COPY r2
```

As we can see, the required order of instruction 0 2 and 3 are still maintained. The semantics of the program has stayed the same.
We can also see that the previously disabled optimization became viable, so the machine copy propagation pass can do the following modification legally:
```llvm
0. r1 = LOAD sp
2. r3 = ADD r1 r4
1. r1 = COPY r5
3. r1 = COPY r2
```

So later on it can be recognized that one instruction is unnecessary (instruction 3) and erase it, so we can end up with more efficient code like this:
```llvm
0. r1 = LOAD sp
2. r3 = ADD r1 r4
1. r1 = COPY r5
```

## Why the current solution is the best
Sub-optimal order of instructions that cause copy propagation optimizations to be missed can appear for many different reasons on almost all the targets.
For this reason, I think the best way to deal with them is in a general target independent analysis of the data dependencies between instructions that are relevant to copy propagation.  

## Limitations
Dependency trees can be complicated. An instruction can have multiple dependencies that can also have multiple dependencies and so on.
Also these dependency trees can have intersections at different points. This extension might be unnecessary since I don't think that there are many cases where we have this complex data dependencies in MIR. (I have only found one in llvm/test/CodeGen/RISCV/rvv/fixed-vectors-interleaved-access-zve32x.ll when checking the MIR during debugging.)

If we want to extend the PR to handle any tree, then the correct order of the instructions in the dependency tree can be calculated with
in order traversal of the dependency graph (and also some merging based on the MI positions in the basic block).

## The current state of the PR
I have tried to check all the tests, but I could not decide if the code is actually correct everywhere. I think in the most places it is correct. I would be really glad if you could also check the tests.

There are some regression tests that are failing. These are hand written tests, whose results can not be generated by update llc or mir scripts. If you see a chance that this PR may get approved and merged I will fix those tests.

## Conclusion
I would like to ask your opinion on this topic. Is it a good direction?
I would be really happy to finish this work, since I enjoy creating stuff like this, if you think it is good direction and could be merged when it is in consistent state.

Do you have another solution in mind to optimize enable optimization in those cases, where unnecessary data dependencies are blocking it?

If you think that this approach is good and I shall continue I will write tests for it and also fix the hand written tests.

Thank you for checking the PR and giving me feedback.

### CODE REVIEW

**Review by qcolombet — CHANGES_REQUESTED** [review]
I think this is the kind of optimization that belongs to postRA scheduling.

I feel this is too complicated for MachineCopyPropagation. Or put differently, until now, MCP was not moving things around so I'm not sure this is the right place to do this.

**Comment by spaits** [comment]
@qcolombet Thank you for your response.

I put this "feature" here because it relies on the already implemented functionality of the pass. This functionality is the recognition of possible optimization in the backward propagation phase. I have just implemented a dependency representation and handling mechanism.

I am open to move this logic to elsewhere but then the full functionality of backward propagation must be also copied there, to recognize the cases when this moving is needed. (Except the register changing and deleting of redundant instructions.)

I was also thinking about if it is a good thing to move around instruction in MCP. My justification was that in this pass we already rename registers and delete instructions. These moves are only happening to enable that. So there can not be any case when the resulting MIR coming out of MCP just has instructions in a different order, moved around. When moving happens then also renaming and deletion will happen.

If there will be others that also think that this logic should be moved to  postRA scheduling, then I will move the dependency handling logic and also do some backward propagation like thing there to identify the cases that needs instruction moving.

**Comment by s-barannikov** [comment]
I agree with @qcolombet
This looks like a job for [dependency breaker](https://github.com/llvm/llvm-project/blob/97324f6274184e607fa6d6cffb1aebee317d4644/llvm/lib/CodeGen/AggressiveAntiDepBreaker.h#L145), which I believe the "new" machine scheduler still doesn't use.

**Comment by spaits** [comment]
So basically this dependency breaker already implements exactly what I have wrote here, but it isn't yet used?

**Comment by s-barannikov** [comment]
> So basically this dependency breaker already implements exactly what I have wrote here, but it isn't yet used?

I'm not quite sure in either part of the statement. We did use the dependency breaker with the new machine scheduler a few years ago in a downstream target, and it did eliminate some copies.

> Or it is rather that, there is this generic dependency breaking stuff that has to be extended to also find and resolve copy propagation related dependencies?

From my recollection, it wasn't too difficult to attach the dependency breaker to the new scheduler, so it may worth a try, it might already be powerful enough.

**Comment by spaits** [comment]
@qcolombet @s-barannikov  I have tried out postRA scheduling as you have suggested. I could get the pass to work on X86 with the following flags for `llc`: `-post-RA-scheduler -break-anti-dependencies=all`.

I used `smuloi8` in `llvm/test/CodeGen/X86/xmulo.ll` for testing where this patch enables optimization:
https://github.com/llvm/llvm-project/pull/98087/files#diff-4e4004fcba25c450462eaec7e0624e6fb03322ae07c833916cd1196e68c97ddfL86 (Also explained on bottom of the message.)

Here is the behavior of post-RA-scheduler on that function on Godbolt (left with postRA scheduling right without it): https://godbolt.org/z/K9z1TrK54

Also you can see that the data dependency was not introduced by the scheduler this time.

Here post-RA-scheduler just replaces `cl` to `dl`. (see Godbolt or bottom of the message) It does not resolve the dependency to enable copy propagation.
I think the goal of this pass is reorder machine instructions to optimize their execution on the target architecture by **considering the latencies** of instructions and the dependencies between them. Not by considering how we could more effective copy propagation. So to use this pass for copy propagation we will need extra code. And based on my observation that extra code would be less fitting for postRA scheduling.

I think this optimization would be pretty useful by looking at the tests it has improved on many different targets in many different contexts. Also there were other attempts to do similar things. I didn't really check them deeply but they did not move around instructions: https://github.com/llvm/llvm-project/pull/74239

I still would prefer this optimization to be in machine-cp. Copy propagation is already doing some very simple data flow analysis like stuff. What I would do just would make this analysis a bit more advanced. If I would extend the responsibilities of post-RA-scheduler I would have to also recognize possible copy propagation in post-RA-scheduler.

Can I ask why do you think that this optimization would belong to post-RA-scheduler?

(Sorry for the long messages I am just trying to exactly describe my thought process and my first language isn't English :) )

---

Explaining the behavior of  my patch and also machine-cp with my patch.
With my patch this happens:
Here is the assembly generated without my patch and without postRA scheduling:
```asm
smuloi8:
# %bb.0:
  movl %ecx, %eax
  imulb %dl
  seto %cl
  movb %al, (%r8)
  movl %ecx, %eax
  retq
```

It recognizes that the by replacing `movb %al, (%r8)` and `seto %cl` we can do a copy propagation.
```asm
smuloi8:
%bb.0:
  movl %ecx, %eax
  imulb %dl
  movb %al, (%r8)
  seto %al
  retq
```

What post-RA-scheduler with break-anti-dependencies=all gives:
```asm
smuloi8:                                # @smuloi8
        mov     eax, ecx
        imul    dl
        seto    dl
        mov     byte ptr [r8], al
        mov     eax, edx
        ret
```
It just does a renaming the dependencies making copy propagation not avalible stay.

(I don't know why there is a difference in the assembly I got form Godbolt and the one I got from my machine from the test generation. The semantics match).
In all cases the `-disable-peephole -mtriple=x86_64-pc-win32` flag was passed to llc.

**Comment by qcolombet** [comment]
I still think it should be part of a scheduling related implementation, because I don't want to recreate the dependency graph in MCP. My worry is that we don't track the right thing and we start breaking things or slowly but surely re-implement the whole DDG in MCP.

I'm surprised that the postRA scheduler doesn't try to get rid of the copies, but maybe this is what you should aim at fixing.

**Comment by spaits** [comment]
@qcolombet Yes I agree, that recreating having the code for data dependency analysis twice is a real possibility that wouldn't be good.

My last concern is this, if we resolve this then I am moving the logic:
- post-RA-scheduler seems kind of not stable to me: No major target uses it on any optimization level (x86, arm, riscv). It outright doesn't run on arm and riscv even if I explicitly ask for the pass with the flag that should enable it. I would like to have this optimization available everywhere after a PR merge, since it is simple and useful.

Is post-RA-scheduler stable? How much work it would be to add it to all the targets from at least O3? Is there any historic reason why it is not used anywhere I don't know of and couldn't find? Is it not done yet? It generates wrong or sub-optimal code? (In my case on X86 it was interesting to me that it has just renamed some regs.)

Also mcp's backward propagation has to be copied into post-RA-scheduler if we move logic there.
Shouldn't we just create a really simple data flow analysis library for MIR (that only operates on basic block)? I would check what exactly happens in post-RA-scheduler and create a library that can be used by post-RA-scheduler and machine-cp or other passes that need data flow analysis like stuff too.

This way we could avoid any duplication of both data flow analysis and copy propagation related code. Also if post-RA-scheduler and dependency breakers are experimental (I deduced this from what I experienced) and will probably stay that for some time then I wouldn't want this to be tied to it, since I think the only relation between my the functionality of my patch and post-RA-scheduler is the need for data flow analysis.

**Comment by s-barannikov** [comment]
> Is post-RA-scheduler stable? How much work it would be to add it to all the targets from at least O3? Is there any historic reason why it is not used anywhere I don't know of and couldn't find? Is it not done yet? It generates wrong or sub-optimal code? (In my case on X86 it was interesting to me that it has just renamed some regs.)

post-RA-scheduler is the legacy scheduler. The "modern" scheduler is machine-scheduler. The legacy scheduler is still used (1) for historical reasons and (2) because it is (arguably?) better suited for DSPs (more accurate resource counting at the cost of compile time). The lack of use of dependency breaker by the new scheduler might be the third reason, but I don't think anyone considered it as a serious disadvantage.

> ...

I once needed tracking functionality used by MCP for a different reason, so in my opinion it is worthwile to make the tracker publicly available.

**Comment by qcolombet** [comment]
> > Is post-RA-scheduler stable? How much work it would be to add it to all the targets from at least O3? Is there any historic reason why it is not used anywhere I don't know of and couldn't find? Is it not done yet? It generates wrong or sub-optimal code? (In my case on X86 it was interesting to me that it has just renamed some regs.)
> 
> post-RA-scheduler is the legacy scheduler. The "modern" scheduler is machine-scheduler. The legacy scheduler is still used (1) for historical reasons and (2) because it is (arguably?) better suited for DSPs (more accurate resource counting at the cost of compile time). The lack of use of dependency breaker by the new scheduler might be the third reason, but I don't think anyone considered it as a serious disadvantage.

Sorry, yes, when I said postRA scheduler, I mean the new one, but running after RA :).

> 
> > ...
> 
> I once needed tracking functionality used by MCP for a different reason, so in my opinion it is worthwile to make the tracker publicly available (regardless of the direction this patch goes).

Sounds like a plan to me.

**Comment by spaits** [comment]
@qcolombet @s-barannikov So what I should do is:
1. Introduce dependency breaking to the new machine-instruction-scheduler
2. Make dependency breaking smarter by taking possible future copy propagation into account

**Comment by spaits** [comment]
@s-barannikov @qcolombet I would like to ask some questions before I start implementing.

So I have looked at the two existing anti dependency breakers. I found out that currently non of these move around instructions to get rid of anti dependencies.

In our case we must move instructions just renaming is not enough. The reason for this is the place final place of data and the origin of the data can not be changed in some scenarios. For example if the values are coming from calls and passed to calls the calling convention determines where that data must be.

For example the above mir's anti dependency can not be resolved by just renaming.
The data returned from `chain` let's call it `$1` must be in `w1` for the next call.
The data loaded from the stack must be in `w0`. So the desired state can be achieved in two instructions, but to do that we must switch the load and the copy.
```mir
BL @chain
renamable $w8 = LDRWui %stack.0.c, 0 # Here we can't load instantly to w0 because it still holds needed data.
renamable $w1 = COPY $w0 # Putting $1 to it's final place for the call.
$w0 = COPY killed renamable $w8
BL @chain
```

```mir
BL @chain
renamable $w1 = COPY $w0 # Here $1 is at it's final place.
renamable $w8 = LDRWui %stack.0.c, 0 # No anti dep. Now machine-cp can optimize.
$w0 = COPY killed renamable $w8
BL @chain
```

(1) Is it ok if from now on it will also move around instructions?

This dependency breaking stuff is pretty huge. (2) Wouldn't it deserve it's own pass?

Also currently post RA scheduling only happens after machine-cp on arm64 and x86 and doesn't happens at all on riscv. (3)Should I add it for riscv? (4)Is it ok if I add another machine-cp after post RA scheduling?

**Comment by spaits** [comment]
Gentle ping @s-barannikov @qcolombet @arsenm . Sorry to ping you. I just want to get feedback for my plan. I want to know if my approach is correct. (I should have done this before this PR to, but instead I just went along with what I thought would be the correct approach from seeing https://discourse.llvm.org/t/how-to-copy-propagate-physical-register-introduced-before-ra/74828 ).

The reason why I want an early opinion from you is:
- It may be bigger work to integrate the existing anti dependency breaker into the new scheduler. Also I want it to be turned on by default when using O3. Is it okay to use anti dep breaker by deafult at some OP level? Which anti dep breaker at which level? (There are two Aggressive or Critical). I would prefer aggressive at O3.
- Currently non of the anti dependency breakers are capable to resolve the dependencies I am complaining about. I must extend one of the two dep breaker implementations. Which one should I extend? Aggressive or Critical? I was suggested aggressive.
- The extension I am planning would involve moving around instructions? Is it fine if the anti dep breaker does that. Currently it just renames.
- Currently the new 
- I want this to work on RISCV. Currently RISCV doesn't have the new post RA scheduler enabled. What should I work on for it to be enabled? (Maybe @topperc could you give any input on this please?)

Sorry for the long message again. I just want to know if what I am planning to do makes sense or is reasonable before I start doing it.

I have also created a discourse post: https://discourse.llvm.org/t/anti-dependency-breaking-and-the-new-sceduler/80109.

Thank you for your time and effort!

**Comment by topperc** [comment]
> I want this to work on RISCV. Currently RISCV doesn't have the new post RA scheduler enabled. What should I work on for it to be enabled? (Maybe @topperc could you give any input on this please?)

It should already be enabled when compiling for any CPU in RISCVProcessors.td that has `FeaturePostRAScheduler`

**Comment by michaelmaitland** [comment]
Note: In this comment, when I say post-RA-scheduler, I mean the PostMachineScheduler.

>  in the cases where an unnecessary data dependency(ies) is(are) blocking an optimization it is recognized and resolved by moving the instruction(s) without changing the semantics
> Can I ask why do you think that this optimization would belong to post-RA-scheduler

In the first statement, you describe the problem as needing to move instructions, but cannot due to unnecessary data dependencies. If this is the problem, then the scheduler is the optimization to do this since it is responsible for reordering instructions limited by data-dependencies.

But as I read on, I started to question whether this really has to do with scheduling. Your end goal is to eliminate unnecessary copies. Coming into the scheduler, there are copies which could have been removed because they could have been propagated. 

If I understand correctly, I think people are saying that the post-RA-scheduler should be responsible for moving instructions so that the MachineCopyPropagation pass can recognize that the copy can be propagated. What you are proposing is that MachineCopyPropagation moves the instructions so that it can do the best copy propagation it can.

I can see both sides. The post-RA-scheduler is balancing many things. To name a few: (1) avoid the number of stalls due to latency (i.e. dependency not ready), (2) avoid the number of stalls due to resource consumption (resources not available), (3) reducing number of spills/reloads. With the proposal that the post-ra scheduler also order instructions to enable copy propagation, we give it yet another thing it needs to balance. There are two concerns I have with using the post-ra scheduler to enable copy propagation:
1. Scheduling for latency, resource consumption, and reducing spills/reloads may not be effective if scheduling for MCP takes precedence.
2. Scheduling for MCP may not be effective if scheduling for latency, resource consumption, and reducing spills/reloads takes precedence.

So by reordering inside of MCP, we can be sure that we reorder to have the best MCP and allow the post-ra scheduler to reorder for the best latency/resource consumption/spills/reloads.

On the other hand, I agree with others when they raise the worry about recreating the dependency graph in MCP. We'd really like to avoid that. Instead of recreating that infrastructure, I wonder about adapting or reusing the ScheduleDAG in MCP. With this approach, I thought about the problem of rebuilding the graph multiple times. I'm not sure how expensive it is, but if we decided that it was too expensive, I wonder if it would be possible to reuse the same ScheduleDAG across multiple passes (i.e. build once and reuse). I think are building the graph multiple times today: once for pre-ra scheduling and once for post-ra scheduling. So that makes me believe that, at least initially, it might be okay to rebuild it a third time.

It sounds like the dependency breaker might be able to solve the problem of reordering in the scheduler to enable MCP, if we decide to take the route that reordering should be done in the scheduler. I wonder how balancing reordering in scheduler for MCP + other things will impact existing scheduling.

If you could reuse the ScheduleDAG infrastructure in MCP, I have no strong objections to allowing MCP to reorder instructions as it wishes. It sounds like we would need consensus around this idea though.

Also, I found a small nit in your explanation:

> Instruction 3 has a dependency on instruction 2 since instruction two uses a value that is in r1 and r1 is re-defined by 3. This means that instruction 2 must preceed instruction 3.
> instruction 2 must preceed instruction 4.

Could you please fix your numbering of the instructions? I think your explanation is sometimes referring to instructions as 1,2,3,4 and sometimes as 0,2,3,4. For example you say instruction 3 has a dependency on instruction 2 when I think you mean instruction 3 has a dependency on instruction 2. Also you refer to instruction 4 when there is no instruction 4 in the 0,1,2,3 based list.

**Comment by spaits** [comment]
@michaelmaitland Thank you very much for your response. I will address your review in more detail tonight or tomorrow morning (Central European Time).

I have looked at the explanation. I fixed the non existent instruction number. I have also took another deeper look at the explanation and added the type of dependency (true, anti and out) to each sentence (at least I hope each sentence :) ) in which I talk about dependencies.

I am sorry for the spelling in my description I have fixed that.

**Comment by spaits** [comment]
> There are two concerns I have with using the post-ra scheduler to enable copy propagation:
> 
>     1. Scheduling for latency, resource consumption, and reducing spills/reloads may not be effective if scheduling for MCP takes precedence.
> 
>     2. Scheduling for MCP may not be effective if scheduling for latency, resource consumption, and reducing spills/reloads takes precedence.
>

Yes. In the case of the C code I show in the issue description, basically the instruction scheduler would "fight" this mechanism that would do the re-ordering for the anti dependency breaker.
- The scheduler would want to prioritize the load.
- The copy pre-copy-propagation re-ordering mechanism I implement would want to keep the load at it's place because it wants to enable that optimization later.  

> On the other hand, I agree with others when they raise the worry about recreating the dependency graph in MCP. 

In my current implementation I have basically created a dumper scheduler graph, that only has the things need to create "must precede" relations between instructions. It doesn't care what kinds of dependencies cause the relation. I know that it is not good practice I also agree that regardless of the direction, the scheduler dag would be the best to re-use. I think the scheduler DAG has everything and more I need to do this optimization. (Basically it needs two things: recognize data dependencies, and be able to update itself when I move around instructions and the order of instructions change.)

> It sounds like the dependency breaker might be able to solve the problem of reordering in the scheduler to enable MCP, if we decide to take the route that reordering should be done in the scheduler.

@s-barannikov you have mentioned anti dep breaker first. Thinking about this deeper, I figured that even if we decide to not do this in MCP maybe it won't be that trivial to find a new place for this in the scheduler:

We are not breaking any anti dependency here. The anti dependencies stay, they must stay, in some cases they are not preventable. Anti dependencies must happen when we have a value, that must be in a specific register (for example the calling convention says so) or must come from a specific register (the calling convention specifies the return value storing register).

For example in the case of the aarch64 example even in the optimal code we must have anti dependency:
```asm
bl      _Z5chainii
mov     w1, w0
ldr     w0, [sp, 28]
bl      _Z5chainii
```

For the optimal work of the MCP we must move around instructions, that have are present in an anti dependency relation with a copy propagation "destination", but do not have any dependency with any of the instructions between itself and a "source" of a copy propagated. (It is pretty weird to write dependency with the `with` preposition but for some reason this describes the situation the best :) .)

So basically we just move around the anti dependency so it won't spoil a possible copy propagation.

> It sounds like we would need consensus around this idea though.

I totally agree.

@michaelmaitland thank you very much checking out this PR/discussion.

**Comment by spaits** [comment]
I think that the logic I want to introduce here is orthogonal to the one that is already present in the anti dependency breaker. We couldn't reuse any logic from the anti dependency breaker. The anti dependencies are found and categorized by the ScheduleDAG. The only thing common between these two is that we both would check if a dependency is anti dependency:

> We are not breaking any anti dependency here. The anti dependencies stay, they must stay, in some cases they are not preventable. Anti dependencies must happen when we have a value, that must be in a specific register (for example the calling convention says so) or must come from a specific register (the calling convention specifies the return value storing register).

I think the best possible solution would be to use the existing scheduler DAG in the MCP. As @michaelmaitland has suggested:

> I wonder about adapting or reusing the ScheduleDAG in MCP. With this approach, I thought about the problem of rebuilding the graph multiple times. I'm not sure how expensive it is, but if we decided that it was too expensive, I wonder if it would be possible to reuse the same ScheduleDAG across multiple passes (i.e. build once and reuse). I think are building the graph multiple times today: once for pre-ra scheduling and once for post-ra scheduling. So that makes me believe that, at least initially, it might be okay to rebuild it a third time.

Also I don't think that the schedule dag building would cost that much. It may be built twice in the legacy post RA scheduler.

What I would do first is removing the dependency graph made by me.

I would use a `scheduleDAG` instead. So the re-implementation of the schedule graph would not happen.
By this I would address this concern of @qcolombet :

> My worry is that we don't track the right thing and we start breaking things or slowly but surely re-implement the whole DDG in MCP.

I will do a more in depth research to see whether any current variant of ScheduleDAG can handle the modification of the underlying MIR. If not then I may create a new subclass that does just that, or I will rebuild the ScheduleDAG at each instruction reordering.

Regardless of this I am still going to work on bringing the anti dependency breaker to the new scheduler. (I have already played around with it with less success).

@s-barannikov @qcolombet @michaelmaitland So what do you think of an implementation in MCP using the existing ScheduleDAG? To me it would be the most logical and we would get the most code reuse this way.

Thank you very much for your attention and response in advance.

**Comment by qcolombet** [comment]
> So what do you think of an implementation in MCP using the existing ScheduleDAG? 

That may work, but be careful about the compile time aspect of such approach. The impact needs to be negligible unless we show significant performance boost across the board (which I doubt we'll achieve)

**Comment by spaits** [comment]
A similar patch like this caused 0.2% reduction on the size for coremark. This was that patch: https://github.com/llvm/llvm-project/pull/73924 . I don't know how big of an impact is that.

**Comment by michaelmaitland** [comment]
> So what do you think of an implementation in MCP using the existing ScheduleDAG?

I think this makes sense. Once you have an implementation, I am happy to take a look into compile time vs performance impact on spec using SiFive hardware.

> The impact needs to be negligible unless we show significant performance boost across the board (which I doubt we'll achieve)

@qcolombet is your main concern the compile time impact or that there will be a lack of performance impact?

**Comment by qcolombet** [comment]
> @qcolombet is your main concern the compile time impact or that there will be a lack of performance impact?

Compile time, or more precisely, it is okay to burn compile time if we get significant performance impact, otherwise the compile time impact needs to be kept at minimum.
